test/test_sparse.py

```diff
@@ -2363,6 +2363,304 @@ def test_sparse_to_numpy(self):
         t = torch.sparse_coo_tensor(torch.tensor(([0, 0], [2, 0])), torch.tensor([1, 4]))
         self.assertRaises(TypeError, lambda: t.numpy())
 
+    def test_softmax(self):
+        if self.device != 'cpu':
+            self.skipTest('sparse softmax not available on %s' % (self.device))
+
+        import torch.nn.functional as F
+
+        def to_dense(sparse, fill_value=None):
+            """
+            Return dense tensor from a sparse tensor using given fill value.
+            """
+            if isinstance(fill_value, str):
+                fill_value = float(fill_value)
+            if fill_value is None or fill_value == 0:
+                return sparse.to_dense()
+            sparse = sparse.coalesce()
+            dense = torch.empty(sparse.shape, dtype=sparse.dtype, device=sparse.device)
+            dense[:] = fill_value
+            for idx, value in zip(sparse._indices().t(), sparse._values()):
+                dense[tuple(idx)] = value
+            return dense
+
+        def softmax_to_dense(sparse, dim):
+            """Dense softmax of a sparse tensor. Useful only for testing softmax
+            correctness.
+
+            When computing softmax of a sparse tensor, the value of
+            unspecified items is negative infinity rather than zero so
+            that
+
+              softmax(sparse.to_dense(fill_value=-inf), dim) == softmax(sparse, dim).to_dense() 
+
+            holds for non-empty lines. One empty lines, the softmax
+            values are defined as 0 in order to preserve the sparsity
+            of result.
+
+            Note that in PyTorch, ``to_dense`` method does not
+            implement the ``fill_value`` keyword argument.
+            """
+            dtype = sparse.dtype
+            device = sparse.device
+            sparse = sparse.coalesce()
```

Comment:
I don't think this coalesce should be necessary

Reply:
Right, `to_dense` calls `coalesce`. Fixed.